src/tools/schema2kotlin.ts

```diff
@@ -39,27 +39,35 @@ export interface KotlinTypeInfo {
   schemaType: string;
 }
 
-function getTypeInfo(name: string, refClassName?: string, refSchemaHash?: string): KotlinTypeInfo {
+function getTypeInfo(opts: {name: string, isCollection?: boolean, refClassName?: string, refSchemaHash?: string}): KotlinTypeInfo {
   const typeMap: Dictionary<KotlinTypeInfo> = {
     'Text': {type: 'String',  decodeFn: 'decodeText()', defaultVal: `""`, schemaType: 'FieldType.Text'},
     'URL': {type: 'String',  decodeFn: 'decodeText()', defaultVal: `""`, schemaType: 'FieldType.Text'},
     'Number': {type: 'Double',  decodeFn: 'decodeNum()',  defaultVal: '0.0', schemaType: 'FieldType.Number'},
     'Boolean': {type: 'Boolean', decodeFn: 'decodeBool()', defaultVal: 'false', schemaType: 'FieldType.Boolean'},
     'Reference': {
-      type: `Reference<${refClassName}>?`,
+      type: `Reference<${opts.refClassName}>`,
       decodeFn: null,
       defaultVal: 'null',
-      schemaType: `FieldType.EntityRef(${quote(refSchemaHash)})`,
+      schemaType: `FieldType.EntityRef(${quote(opts.refSchemaHash)})`,
     },
   };
 
-  const info = typeMap[name];
+  const info = typeMap[opts.name];
   if (!info) {
-    throw new Error(`Unhandled type '${name}' for kotlin.`);
+    throw new Error(`Unhandled type '${opts.name}' for kotlin.`);
   }
-  if (name === 'Reference') {
-    assert(refClassName, 'refClassName must be provided for References');
-    assert(refSchemaHash, 'refSchemaHash must be provided for References');
+  if (opts.name === 'Reference') {
+    assert(opts.refClassName, 'refClassName must be provided for References');
+    assert(opts.refSchemaHash, 'refSchemaHash must be provided for References');
+    if (!opts.isCollection) {
+      // Singleton Reference fields are nullable.
+      info.type += '?';
```

Comment:
Did anyone ever discuss the alternate, a Reference that always returns null when you deref it? Or as a Collection with only 0 or 1 refs? I suppose in Kotlin with first class nullability support, boxed types are frowned upon (e.g. Optional/Maybe types), but it does allow you do write code that map()s over fields without having to sprinkle null-safe operators.

Reply:
It hasn't come up so far. But Kotlin makes dealing with nulls so easy that I'm not sure it's worth it tbh.